### PR TITLE
feat: added http.route in rails instrumentation to match the spec

### PR DESCRIPTION
**Pull request by angeliski**
# Description
The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions) says:

> Given an inbound request for a route (e.g. "/users/:userID?") the name attribute of the span SHOULD be set to this route. If the route does not include the application root, it SHOULD be prepended to the span name.
> 
> If the route cannot be determined, the name attribute MUST be set as defined in the general semantic conventions for HTTP.

Some exporters need more information to provide useful insights to the user,
so add `http.route` could make more easily convert traces to the final destination.

My current case is using Datadog: The [exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/translate_traces.go#L440) uses the `http.route` to define how the resource will be shown, so I think makes sense to add here that information.

I created a really simple implementation, so any feedback would be useful 

The result is something like that:

Screenshot from 2021-03-17 16-20-46

### CODE REVIEW

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb
Comment (robertlaurin) [inline instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:18]: When I originally implemented this instrumentation I had some concerns regarding the overhead of running [.recognize](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/journey/router.rb#L65) a second time for each request to the application.  

I'd like to get a second reviewer to weigh in on it, and if they don't share my same concerns then I have no reason to block on this.

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb
Comment (robertlaurin) [inline instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:16]: The reason we opted for this approach is that we believe using the controller name and action as the span name is more conventional in the context of a rails application.

As you pointed out this is not following the spec, would it make more sense for your use case if we added a configuration that allowed you to set the span name to the path template instead?

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb
Comment (angeliski) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:18]: Yes, I can understand your concern, I didn't find a better way to get the route here. Maybe we can add a configuration to enable that behavior?

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb
Comment (angeliski) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:16]: More or less, the spec is a little confused about that.

About the [name](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span)
> The span name concisely identifies the work represented by the Span, for example, an RPC method name, a function name, or the name of a subtask or stage within a larger computation. The span name SHOULD be the most general string that identifies a (statistically) interesting class of Spans, rather than individual Span instances while still being human-readable. That is, "get_user" is a reasonable name, while "get_user/314159", where "314159" is a user ID, is not a good name due to its high cardinality. Generality SHOULD be prioritized over human-readability.

So in that case make sense to me, in a rails app, it behavior you used.
But we have a conflict when we start to talk about the [HTTP Server semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions)

> Given an inbound request for a route (e.g. "/users/:userID?") the name attribute of the span SHOULD be set to this route. If the route does not include the application root, it SHOULD be prepended to the span name.
>
> If the route cannot be determined, the name attribute MUST be set as defined in the general semantic conventions for HTTP.

In that case, the span name AND route should be the same? That doesn't make sense to me (could be a problem in specification not here).

The problem (to me) in only change the span name is the current behavior in [Datadog exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/translate_traces.go#L441) they use the route to build the resource (and show grouped in UI). I only thought of adding a route here, but maybe we need to change the exporter behavior to look to other attributes.

Comment (angeliski) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:16]: Hey, @ericmustin I see you have some commits in Datadog exporter, could you help us here in that discussion?
What do you think is the better approach?

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb
Comment (fbogsany) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:18]: 👍 I'd be pro-config-option

Comment (robertlaurin) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:18]: Agreed, if we put this behind a config option that defaults to it being off I think that's a good compromise for bringing this functionality in.

Comment (angeliski) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:18]: Great! I will work on it

Comment (angeliski) [reply instrumentation/rails/lib/opentelemetry/instrumentation/rails/patches/action_controller/metal.rb:25]: Hey team, I updated the code, please take a look

File: instrumentation/rails/lib/opentelemetry/instrumentation/rails/instrumentation.rb
Comment (robertlaurin) [inline instrumentation/rails/lib/opentelemetry/instrumentation/rails/instrumentation.rb:25]: I went back and forth on wether or not this was the right name for this setting, typically I wouldn't want to associate the name with the implementation (ie. `router.recognize`).  But in the case I think this makes sense as the reason we are putting this behind a config is due to performance concerns.

File: instrumentation/rails/README.md
Comment (robertlaurin) [inline instrumentation/rails/README.md:36]: Instead of pointing to the main branch which may lose context as it changes, lets just use the current latest tag.
```suggestion
The `http.route` attribute is disabled by default because we use [.recognize](https://github.com/rails/rails/blob/v6.1.3/actionpack/lib/action_dispatch/journey/router.rb#L65)
```

**Review by robertlaurin — APPROVED** [review]
I left one suggested change around the readme that I think should come in before merge, otherwise this looks good. 

Thank you.

**Review by fbogsany — APPROVED** [review]

**Pull request merged**
